File: yams/cpp/src/yams/motorcontrollers/SmartMotorControllerCommandRegistry.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

#include "yams/motorcontrollers/SmartMotorControllerCommandRegistry.hpp"

#include <wpi/smartdashboard/SmartDashboard.hpp>
#include <wpi/commands2/Commands.hpp>

#include <stdexcept>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

namespace yams::motorcontrollers {

std::unordered_map<std::string, wpi::cmd::CommandPtr> SmartMotorControllerCommandRegistry::s_commands;
std::unordered_map<std::string, std::vector<std::function<void()>>>
    SmartMotorControllerCommandRegistry::s_callbacks;
std::unordered_map<std::string, wpi::cmd::SubsystemBase*> SmartMotorControllerCommandRegistry::s_owners;

std::string SmartMotorControllerCommandRegistry::MakeKey(const std::string& cmdName,
                                                         wpi::cmd::SubsystemBase* subsystem) {
  return subsystem->GetName() + "/" + cmdName;
}

void SmartMotorControllerCommandRegistry::PublishToNT(const std::string& cmdName,
                                                      wpi::cmd::SubsystemBase* subsystem) {
  auto key = MakeKey(cmdName, subsystem);
  s_commands.insert_or_assign(key, wpi::cmd::Run(
                                       [capturedKey = key] {
                                         for (auto& cb : s_callbacks[capturedKey]) cb();
                                       },
                                       {subsystem})
                                       .WithName(cmdName));
  wpi::SmartDashboard::PutData("Mechanisms/Commands/" + key, s_commands.at(key).get());
}

void SmartMotorControllerCommandRegistry::AddCommand(const std::string& cmdName,
                                                     wpi::cmd::SubsystemBase* subsystem,
                                                     std::function<void()> callback) {
  auto key = MakeKey(cmdName, subsystem);
  auto ownerIt = s_owners.find(key);
  if (ownerIt != s_owners.end() && ownerIt->second != subsystem) {
    throw std::runtime_error("SmartMotorControllerCommandRegistry: subsystem name conflict — \"" +
                             subsystem->GetName() +
                             "\" is already registered by a different subsystem instance. "
                             "Use unique subsystem names for each subsystem instance (e.g. "
                             "\"LeftTurret\", \"RightTurret\").");
  }
  s_owners[key] = subsystem;
  s_callbacks[key].push_back(std::move(callback));
  if (!CommandExists(cmdName, subsystem)) {
    PublishToNT(cmdName, subsystem);
  }
}

bool SmartMotorControllerCommandRegistry::CommandExists(const std::string& cmdName,
                                                        wpi::cmd::SubsystemBase* subsystem) {
  return s_commands.count(MakeKey(cmdName, subsystem)) > 0;
}

void SmartMotorControllerCommandRegistry::RemoveCommands(wpi::cmd::SubsystemBase* subsystem) {
  for (auto it = s_owners.begin(); it != s_owners.end();) {
    if (it->second == subsystem) {
      s_commands.erase(it->first);
      s_callbacks.erase(it->first);
      it = s_owners.erase(it);
    } else {
      ++it;
    }
  }
}

void SmartMotorControllerCommandRegistry::Clear() {
  s_commands.clear();
  s_callbacks.clear();
  s_owners.clear();
}

}  // namespace yams::motorcontrollers

File: yams/java/yams/motorcontrollers/SmartMotorControllerCommandRegistry.java
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

package yams.motorcontrollers;

import org.wpilib.smartdashboard.SmartDashboard;
import org.wpilib.command2.Command;
import org.wpilib.command2.Commands;
import org.wpilib.command2.Subsystem;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Registry for WPILib {@link org.wpilib.command2.Command} instances associated with a
 * {@link yams.motorcontrollers.SmartMotorController}. Mechanisms register setpoint commands here
 * so the scheduler can manage them.
 */
public class SmartMotorControllerCommandRegistry
{
  /**
   * HashMap with the Subsystem name as the key and the shared command which runs all runnables added to the subsystem.
   */
  private static Map<String, Command>        commands         = new HashMap<>();
  /**
   * HashMap with the Subsystem name as the key and a list of runnables to be added to the shared command.
   */
  private static Map<String, List<Runnable>> commandCallbacks = new HashMap<>();
  /**
   * HashMap tracking which subsystem instance owns each command key.
   * Used to detect when two different subsystem instances share the same name.
   */
  private static Map<String, Subsystem>      commandOwners    = new HashMap<>();

  /**
   * Create the {@link Command} and publish it to NetworkTables.
   *
   * @param cmdName   Command name to publish to NetworkTables.
   * @param subsystem Subsystem to create the command for.
   */
  private static void addCommandToNT(String cmdName, Subsystem subsystem)
  {
    var key = subsystem.getName() + "/" + cmdName;
    Command cmd = Commands.run(() -> {
      for (var callback : commandCallbacks.get(key))
      {
        callback.run();
      }
    }, subsystem).withName(cmdName);
    commands.put(key, cmd);
    SmartDashboard.putData("Mechanisms/Commands/" + key, cmd);
  }

  /**
   * Add a command to the registry.
   *
   * @param cmdName   Command name to publish to NetworkTables.
   * @param subsystem Subsystem to create the command for.
   * @param callback  Runnable to be added to the shared command.
   */
  public static void addCommand(String cmdName, Subsystem subsystem, Runnable callback)
  {
    var key   = subsystem.getName() + "/" + cmdName;
    var owner = commandOwners.get(key);
    if (owner != null && owner != subsystem)
    {
      throw new IllegalStateException(
          "SmartMotorControllerCommandRegistry: subsystem name conflict — \"" +
          subsystem.getName() + "\" is already registered by a different subsystem instance. " +
          "Use unique subsystem names for each subsystem instance (e.g. \"LeftTurret\", \"RightTurret\").");
    }
    commandOwners.put(key, subsystem);
    commandCallbacks.computeIfAbsent(key, k -> new ArrayList<>()).add(callback);
    // Create Command and publish it to NT
    if (!commandExists(cmdName, subsystem))
    {addCommandToNT(cmdName, subsystem);}
  }

  /**
   * Check if a command exists.
   *
   * @param cmdName   Command name.
   * @param subsystem Subsystem.
   * @return True if command exists.
   */
  public static boolean commandExists(String cmdName, Subsystem subsystem)
  {
    var key = subsystem.getName() + "/" + cmdName;
    return commands.containsKey(key);
  }

  /**
   * Remove all commands registered for the given subsystem instance.
   *
   * <p>Call this when a subsystem is being torn down (e.g. between tests) so that a
   * new instance with the same name can register without triggering the conflict check.
   *
   * @param subsystem Subsystem whose commands should be removed.
   */
  public static void removeCommands(Subsystem subsystem)
  {
    commandOwners.entrySet().removeIf(e -> {
      if (e.getValue() == subsystem)
      {
        commandCallbacks.remove(e.getKey());
        commands.remove(e.getKey());
        return true;
      }
      return false;
    });
  }

}
